### ISSUE TITLE
API Proposal: SequenceReader<T>.ClearCache() method to allow the reader to re-use previously allocated memory

### ISSUE THREAD

**Issue by softworkz** [dotnet/runtime#31067]
## Purpose
Provide a method that allows consumers of the SequenceReader<T> to indicate that it is no longer interested in the previously returned results and that the SequenceReader<T> may re-use memory it has internally allocated before.

```cs
public partial ref struct SequenceReader<T>
{
    void ClearCache();
}
```
## Background
There are cases where the SequenceReader needs to allocate memory in order to return results because the requested data is crossing segment boundaries and cannot be returned as a contiguous range of memory otherwise.
In the context of https://github.com/dotnet/corefx/issues/40962, additional SequenceReader methods will be added where this can happen.
The reasons why this needs to happen and why it cannot be avoided in order to make reasonable use of the SequenceReader<T> are already discussed in various other issues.
Those APIs can be used in reasonable ways with low rates of extra-allocation and in less reasonable ways (see here https://github.com/dotnet/corefx/issues/40871#issuecomment-530088885 for some figures regarding allocation probability depending on the length of requested data)

## Example
Consider the following parsing loop:
```cs
void Parse(ref ReadOnlySequence<byte> buffer)
{
    var reader = new SequenceReader<byte>(buffer);

    while (!reader.End)
    {
        if (!reader.TryRead(8, out ReadOnlySpan<byte> payload)))
        {
			return;
        }

        // Do something with those 8 payload bytes
    }
}
```

## Problem
Even in very reasonable use cases like the above, there's one problem: The allocations that the SequenceReader<T> needs to make, are adding up over time as the SequenceReader<T> needs to make sure that all returned results remain valid during the readers lifetime as it cannot make any assumptions about the calling code's interest in results returned previously (in the example above: it doesn't know that there's a loop and and the calling code looses interest in earlier result with each iteration).

## Solution
We need to provide a method that allows the calling code to tell the SequenceReader<T> that it is no longer interested in earlier results.
This in turn will allow the internal implementation of the SequenceReader<T> to re-use previously allocated memory for future results - which makes that memory become a kind of "cache".

Hence this API proposal for a **ClearCache()** method.

**Comment by GrabYourPitchforks** [comment]
cc @JeremyKuhne

**Comment by JeremyKuhne** [comment]
I don't think we should add allocation management to `SequenceReader`. I think the right answer is to provide an `IAllocator` interface that `SequenceReader` can use that will allow plugging in arbitrary allocation strategies. Something like:

``` C#
    public interface IAllocator<T>
    {
        // Maybe T[]? Memory<T> has negative performance implications
        public Span<T> Allocate(int minimumLength);
    }

    // What SequenceReader already does
    public class ArrayAllocator<T> : IAllocator<T>
    {
        public Span<T> Allocate(int minimumLength) => new T[minimumLength];
    }

    // I think this satisfies what is asked in this issue?
    public class ArrayPoolAllocator<T> : IAllocator<T>
    {
        private Stack<T[]> _rents = new Stack<T[]>();

        public Span<T> Allocate(int minimumLength)
        {
            T[] array = ArrayPool<T>.Shared.Rent(minimumLength);
            _rents.Push(array);
            return array;
        }

        public void ReturnBuffers()
        {
            while (_rents.TryPop(out T[] array))
            {
                ArrayPool<T>.Shared.Return(array);
            }
        }
    }
```
